File: ios/RideGuard/Core/Update/UpdateManifest.swift
import Foundation

//  The shared update manifest, published at a stable raw URL on the GitHub
//  repo and read by BOTH platforms:
//
//      https://raw.githubusercontent.com/zigazaga4/rideguard/main/updates/latest.json
//
//  Android downloads an APK from it and installs it with `PackageInstaller`.
//  iOS cannot install anything from inside an app, so it uses the `manifestUrl`
//  to hand iOS an `itms-services://` OTA install — see `docs/ios-updates.md`.
//  The file format is identical on purpose: one release process, one file to
//  get right, and a version skew between the two apps is immediately visible.

/// One release, as published to GitHub.
///
/// Decoding is hand-written rather than synthesised because every rule this
/// type has to honour is a leniency rule: a missing `mandatory`, a missing
/// platform block, or a malformed sibling block must all leave the rest of the
/// manifest usable. Synthesised `Codable` throws on the first missing key,
/// which would turn "no Android build this time" into "update check broken".
public struct UpdateManifest: Equatable, Sendable {

    /// The only schema this build understands. Anything else — higher, lower,
    /// or nonsense — means the payload may not mean what we think it means.
    public static let supportedSchemaVersion = 1

    public let schemaVersion: Int
    /// ISO-8601, kept as the raw string. Parsed on demand rather than through
    /// a `JSONDecoder` date strategy, so a publisher who writes a slightly odd
    /// timestamp costs us a formatted date, not the whole update check.
    public let publishedAt: String?
    public let notes: String
    public let mandatory: Bool
    public let android: AndroidRelease?
    public let ios: IOSRelease?

    public init(
        schemaVersion: Int,
        publishedAt: String? = nil,
        notes: String = "",
        mandatory: Bool = false,
        android: AndroidRelease? = nil,
        ios: IOSRelease? = nil
    ) {
        self.schemaVersion = schemaVersion
        self.publishedAt = publishedAt
        self.notes = notes
        self.mandatory = mandatory
        self.android = android
        self.ios = ios
    }

    public var publishedDate: Date? {
        guard let publishedAt else { return nil }
        return ISO8601DateFormatter().date(from: publishedAt)
    }

    /// The Android half. iOS never acts on it — it is decoded so a malformed
    /// Android block is proven harmless rather than assumed to be.
    public struct AndroidRelease: Equatable, Sendable {
        public let versionCode: Int
        public let versionName: String
        public let url: String
        public let sha256: String?
        public let sizeBytes: Int64?
        public let minSdk: Int?

        public init(
            versionCode: Int,
            versionName: String,
            url: String,
            sha256: String? = nil,
            sizeBytes: Int64? = nil,
            minSdk: Int? = nil
        ) {
            self.versionCode = versionCode
            self.versionName = versionName
            self.url = url
            self.sha256 = sha256
            self.sizeBytes = sizeBytes
            self.minSdk = minSdk
        }
    }

    public struct IOSRelease: Equatable, Sendable {
        /// Display string. For humans only — never compared.
        public let version: String
        /// `CFBundleVersion`. THE comparison key: an integer that only ever
        /// goes up. Comparing "1.10.0" against "1.9.0" as strings puts the
        /// newer build behind the older one, which is how an updater ends up
        /// silently never updating.
        public let build: Int
        /// HTTPS URL of the `manifest.plist` handed to `itms-services://`.
        /// Apple requires HTTPS with a valid certificate; a GitHub release
        /// asset URL qualifies.
        public let manifestUrl: String
        /// Below this, the OTA install will fail on the device. Optional, and
        /// treated as "no floor" when absent.
        public let minIosVersion: String?
        /// Not in the v1 schema. Decoded when a publisher includes it, because
        /// the install prompt gives the driver no size at all and a driver on
        /// mobile data deserves to know before tapping.
        public let sizeBytes: Int64?

        public init(
            version: String,
            build: Int,
            manifestUrl: String,
            minIosVersion: String? = nil,
            sizeBytes: Int64? = nil
        ) {
            self.version = version
            self.build = build
            self.manifestUrl = manifestUrl
            self.minIosVersion = minIosVersion
            self.sizeBytes = sizeBytes
        }

        /// The URL that makes iOS offer to install this build.
        ///
        /// Nil when the manifest URL is not HTTPS: iOS silently refuses plain
        /// HTTP here, and a button that does nothing when tapped is worse than
        /// a button that is not offered.
        public var otaInstallURL: URL? {
            UpdateManifest.otaInstallURL(forManifestAt: manifestUrl)
        }
    }

    /// `itms-services://?action=download-manifest&url=<https url>`
    ///
    /// The URL is percent-encoded against an explicitly unreserved set rather
    /// than `.urlQueryAllowed`, which leaves `&` and `=` untouched — a release
    /// asset URL carrying a query string would otherwise truncate the
    /// parameter and iOS would report a meaningless "cannot connect".
    public static func otaInstallURL(forManifestAt manifestUrl: String) -> URL? {
        let trimmed = manifestUrl.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let parsed = URL(string: trimmed), parsed.scheme?.lowercased() == "https" else { return nil }
        var unreserved = CharacterSet.alphanumerics
        unreserved.insert(charactersIn: "-._~")
        guard let encoded = trimmed.addingPercentEncoding(withAllowedCharacters: unreserved) else { return nil }
        return URL(string: "itms-services://?action=download-manifest&url=\(encoded)")
    }
}

// MARK: - Decoding

private extension KeyedDecodingContainer {
    /// Absent, null, or the wrong type all read as nil.
    ///
    /// A publisher who writes `"mandatory": "false"` should cost us that one
    /// field, not the whole update check. Everything the updater does is a
    /// convenience; nothing about it is worth failing loudly over.
    func lenient<T: Decodable>(_ type: T.Type, _ key: Key) -> T? {
        (try? decodeIfPresent(type, forKey: key)) ?? nil
    }
}

extension UpdateManifest: Decodable {

    private enum Key: String, CodingKey {
        case schemaVersion, publishedAt, notes, mandatory, android, ios
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: Key.self)
        // The one genuinely required field. Without it we cannot know whether
        // the rest of the payload means what we think it means.
        schemaVersion = try c.decode(Int.self, forKey: .schemaVersion)
        publishedAt = c.lenient(String.self, .publishedAt)
        notes = c.lenient(String.self, .notes) ?? ""
        mandatory = c.lenient(Bool.self, .mandatory) ?? false
        // Each block is decoded leniently on its own: a broken Android block
        // must not cost an iPhone its update, and vice versa.
        android = c.lenient(AndroidRelease.self, .android)
        ios = c.lenient(IOSRelease.self, .ios)
    }
}

extension UpdateManifest.AndroidRelease: Decodable {
    private enum Key: String, CodingKey {
        case versionCode, versionName, url, sha256, sizeBytes, minSdk
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: Key.self)
        versionCode = try c.decode(Int.self, forKey: .versionCode)
        versionName = c.lenient(String.self, .versionName) ?? ""
        url = try c.decode(String.self, forKey: .url)
        sha256 = c.lenient(String.self, .sha256)
        sizeBytes = c.lenient(Int64.self, .sizeBytes)
        minSdk = c.lenient(Int.self, .minSdk)
    }
}

extension UpdateManifest.IOSRelease: Decodable {
    private enum Key: String, CodingKey {
        case version, build, manifestUrl, minIosVersion, sizeBytes
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: Key.self)
        // `build` and `manifestUrl` are the two fields without which this block
        // is not actionable, so their absence makes the block absent — which
        // reads downstream as "no iOS build in this release", never an error.
        version = c.lenient(String.self, .version) ?? ""
        build = try c.decode(Int.self, forKey: .build)
        manifestUrl = try c.decode(String.self, forKey: .manifestUrl)
        minIosVersion = c.lenient(String.self, .minIosVersion)
        sizeBytes = c.lenient(Int64.self, .sizeBytes)
    }
}

// MARK: - The installed build

/// What this copy of the app is, for comparison against the manifest.
public struct AppVersion: Equatable, Sendable {
    /// `CFBundleShortVersionString` — "1.2.0". Shown, never compared.
    public let version: String
    /// `CFBundleVersion` — the monotonic integer everything is decided on.
    public let build: Int

    public init(version: String, build: Int) {
        self.version = version
        self.build = build
    }

    /// A `CFBundleVersion` that is not an integer (say "1.2.0") reads as build
    /// 0, which makes every published build look newer. That is the right
    /// failure direction — the driver is offered an update they may not need,
    /// rather than silently pinned to an old build forever — but the release
    /// script should keep `CFBundleVersion` an integer. See `ios/README.md`.
    public static func current(bundle: Bundle = .main) -> AppVersion {
        current(info: bundle.infoDictionary)
    }

    /// Split out from the `Bundle` accessor so the parsing is testable without
    /// subclassing `Bundle`, which has no public designated initialiser worth
    /// fighting.
    public static func current(info: [String: Any]?) -> AppVersion {
        let info = info ?? [:]
        let version = info["CFBundleShortVersionString"] as? String ?? "0"
        let rawBuild = info["CFBundleVersion"] as? String ?? "0"
        return AppVersion(version: version, build: Int(rawBuild) ?? 0)
    }
}

File: android/domain/src/main/kotlin/com/rideguard/domain/update/UpdateManifest.kt
package com.rideguard.domain.update

import kotlinx.serialization.Serializable

/**
 * The update manifest, published to GitHub and read by both platforms.
 *
 * This app is distributed outside both stores — sideloaded on Android,
 * ad-hoc-signed on iOS — so nothing is going to tap the driver on the
 * shoulder when a new build exists. This file is that shoulder tap. It lives
 * at a stable raw URL on the repo:
 *
 *     https://raw.githubusercontent.com/zigazaga4/rideguard/main/updates/latest.json
 *
 * A raw URL rather than the GitHub Releases API on purpose: no rate limit
 * worth worrying about, no auth, no pagination, and the shape is ours rather
 * than GitHub's, so it cannot change under us.
 *
 * ## Rules that both platforms MUST implement identically
 *
 * 1. Both platform blocks are OPTIONAL. Shipping an Android-only release is
 *    normal, and an iPhone reading that manifest must conclude "nothing for
 *    me", never "error".
 * 2. Compare on the INTEGER ([AndroidRelease.versionCode] / [IosRelease.build]).
 *    Display strings are for humans and sort terribly — "1.10.0" is older than
 *    "1.9.0" under every string comparison ever written.
 * 3. Unknown fields are ignored, so adding one later cannot brick an old
 *    client. That is why the decoder sets `ignoreUnknownKeys`.
 * 4. A [schemaVersion] we do not recognise means WE are too old to be reading
 *    this safely — say so and point at a manual download, rather than guessing
 *    at the payload and installing something unexpected.
 * 5. Nothing here is ever fatal. Failing to reach GitHub must not stop a
 *    driver mid-shift; the updater is a convenience and degrades to silence.
 */
@Serializable
data class UpdateManifest(
    val schemaVersion: Int = 1,
    /** ISO-8601 UTC. Display only — never used for ordering. */
    val publishedAt: String? = null,
    /** Short changelog, shown verbatim. May contain newlines. */
    val notes: String? = null,
    /**
     * A release important enough to nag about. It still does not install
     * itself — see [UpdateStatus.Available].
     */
    val mandatory: Boolean = false,
    val android: AndroidRelease? = null,
    val ios: IosRelease? = null,
)

@Serializable
data class AndroidRelease(
    /** The only field ordering is decided on. */
    val versionCode: Int,
    /** Human-facing, e.g. "1.2.0". */
    val versionName: String,
    /** Direct HTTPS link to the APK — a GitHub release asset. */
    val url: String,
    /**
     * Lowercase hex SHA-256 of the APK.
     *
     * Not paranoia: this is served over plain HTTPS from a public repo and
     * then handed to the package installer. Checking the digest before install
     * is the difference between "we downloaded the right file" and "we
     * downloaded a file". Null means unverifiable, which callers should treat
     * as a reason to warn, not as a reason to trust.
     */
    val sha256: String? = null,
    /** For showing "42 MB" before a driver spends his mobile data on it. */
    val sizeBytes: Long? = null,
    /** Refuse to offer a build the device cannot run. */
    val minSdk: Int = 0,
)

@Serializable
data class IosRelease(
    val version: String,
    /** The only field ordering is decided on. */
    val build: Int,
    /**
     * HTTPS URL of the `itms-services` manifest plist.
     *
     * iOS has no API for an app to install an app, so the OTA plist is the
     * whole mechanism. It only works for ad-hoc/enterprise-signed builds, and
     * both this URL and the .ipa it points at must be HTTPS with a valid
     * certificate — GitHub release assets qualify.
     */
    val manifestUrl: String,
    val minIosVersion: String? = null,
)

/** What the app should do about [UpdateManifest], once compared to itself. */
sealed interface UpdateStatus {

    /** Already current. The common case, and it should be silent. */
    data object UpToDate : UpdateStatus

    /**
     * A newer build exists.
     *
     * Deliberately NOT self-installing. This runs on a working phone during a
     * shift; deciding on the driver's behalf to replace the app while he is
     * waiting for a ride is not a convenience, it is an outage.
     */
    data class Available(
        val release: AndroidRelease,
        val notes: String?,
        val mandatory: Boolean,
    ) : UpdateStatus

    /** The manifest is newer than this build knows how to read. */
    data class TooOldToUnderstand(val manifestSchema: Int) : UpdateStatus

    /** A newer build exists but this device cannot run it. */
    data class Unsupported(val reason: String) : UpdateStatus

    /** Network failure, garbage JSON, whatever. Non-fatal by construction. */
    data class Unreadable(val reason: String) : UpdateStatus
}
